okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -692,6 +696,38 @@ public Builder addEncodedPathSegment(String encodedPathSegment) {
       return this;
     }
 
+    public Builder setPathSegment(int index, String pathSegment) {
+      if (pathSegment == null) throw new IllegalArgumentException("pathSegment == null");
```

Comment:
Missing index bounds check. Or if you are falling back to List to do the work then tests are missing.